lib/rucio/core/oidc.py

```diff
@@ -0,0 +1,924 @@
+# Copyright 2012-2018 CERN for the benefit of the ATLAS collaboration.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Authors:
+# - Jaroslav Guenther <jaroslav.guenther@cern.ch>, 2019
+#
+# PY3K COMPATIBLE
+
+import json
+import random
+import subprocess
+import time
+import traceback
+
+from datetime import datetime, timedelta
+from jwkest.jws import JWS
+from jwkest.jwt import JWT
+from oic import rndstr
+from oic.oic import Client, Token, REQUEST2ENDPOINT
+from oic.oauth2.message import CCAccessTokenRequest
+from oic.oic.message import (AccessTokenResponse, AuthorizationResponse,
+                             Message, RegistrationResponse)
+from oic.utils.authn.client import CLIENT_AUTHN_METHOD
+from rucio.common.config import config_get
+from rucio.common.exception import (CannotAuthenticate, CannotAuthorize,
+                                    RucioException)
+from rucio.common.types import InternalAccount
+from rucio.common.utils import (build_url, filter_thread_load,
+                                oidc_identity_string, query_bunches,
+                                val_to_space_sep_str)
+from rucio.core.account import account_exists
+from rucio.core.identity import exist_identity_account, get_default_account
+from rucio.core.monitor import record_counter, record_timer
+from rucio.db.sqla import models
+from rucio.db.sqla.constants import IdentityType
+from rucio.db.sqla.session import read_session, transactional_session
+
+try:
+    # Python 2
+    from urlparse import urlparse, parse_qs
+except ImportError:
+    # Python 3
+    from urllib.parse import urlparse, parse_qs
+
+# worokaround for a bug in pyoidc (as of Dec 2019)
+REQUEST2ENDPOINT['CCAccessTokenRequest'] = 'token_endpoint'
+
+# private/protected file containing Rucio Client secrets known to the Identity Provider as well
+IDPSECRETS = config_get('oidc', 'idpsecrets', False)
+ADMIN_ISSUER_ID = config_get('oidc', 'admin_issuer', False)
+
+# TO-DO permission layer: if scope == 'wlcg.groups'
+# --> check 'profile' info (requested profile scope)
+
+
+def get_rucio_oidc_clients(keytimeout=43200):
```

Comment:
I didn't fully check all these methods, but some look like they are only used within the module. if they are only used within the module, I would make them internal `_` for better understanding.

Reply:
Done.